data/nat/basic.lean

```diff
@@ -544,4 +576,93 @@ theorem dvd_fact : ∀ {m n}, m > 0 → m ≤ n → m ∣ fact n
 theorem fact_le {m n} (h : m ≤ n) : fact m ≤ fact n :=
 le_of_dvd (fact_pos _) (fact_dvd_fact h)
 
+section find_greatest
+
+variables {P : ℕ → Prop} [decidable_pred P] {bound : ℕ}
+    (hall : ∀ m : ℕ, m > bound → ¬ P m) (hex : ∃ m, P m)
+
+protected def nat.find_greatest_x : {n : ℕ // P n ∧ ∀ m : ℕ, m > n → ¬ P m} :=
+have ∃ v, P (bound - v), from 
+  let ⟨m, Hpm⟩ := hex in
+  have m ≤ bound, from le_of_not_gt (λ h, hall _ h Hpm),
+  ⟨bound - m, by rw nat.sub_sub_self; repeat {assumption}⟩,
+let minv := nat.find this,
+    minv_spec := nat.find_spec this,
+    minv_lt := @nat.find_min _ _ this in 
+⟨ bound - minv,
+  minv_spec,
+  assume (m : ℕ) (h : m > bound - minv),
+  if hmb : m > bound then
+    hall _ hmb
+  else
+    have hmb' : m ≤ bound, from le_of_not_gt hmb,
+    let minv' := @minv_lt (bound - m) in
+    have m + minv > bound, from nat.lt_add_of_sub_lt_right h,
+    have bound - m < minv, from (nat.sub_lt_left_iff_lt_add hmb').2 this,
+    begin rw nat.sub_sub_self at minv', exact minv' this, exact hmb' end ⟩ 
+
+protected def nat.find_greatest : ℕ := nat.find_greatest_x hall hex
```

Comment:
Unlike `nat.find`, this doesn't require any fancy arguments for termination, so I think it should be defined as a total function given `P` and `bound` (I guess it can return an `option` in the core version, since it can determine itself if there is a witness).